From 212bb835047847f0cd7008704028b0fffa2f4911 Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <icenowy@aosc.io>
Date: Wed, 26 Jun 2019 00:44:43 +0800
Subject: [PATCH 3/4] platform/x86: sony-laptop: do ACPI initialization on new
 NC devices

On new VAIO NC devices, the SSDT of them requires the EC region to be
available before the _STA method of them to return available status.

As we probe the _STA of the devices before we make EC region available,
the _INI method is not called, thus the initialization of NC device is
skipped. In these devices, _INI is used to fill the function handle
table, and its skipping will lead sony-laptop to fail to acquire
function handles.

Do the ACPI initialization by ourselves on new NC devices. The existence
of ACPI method SIRN of the NC device can be used to detect whether EC is
needed, because the EC _REG method uses it to notify NC device.

Signed-off-by: Icenowy Zheng <icenowy@aosc.io>
---
 drivers/platform/x86/sony-laptop.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/platform/x86/sony-laptop.c b/drivers/platform/x86/sony-laptop.c
index 459fb23a3e3d..c96211182c05 100644
--- a/drivers/platform/x86/sony-laptop.c
+++ b/drivers/platform/x86/sony-laptop.c
@@ -3209,6 +3209,21 @@ static int sony_nc_add(struct acpi_device *device)
 			dprintk("ECON Method failed\n");
 	}
 
+	if (acpi_has_method(sony_nc_acpi_handle, "SIRN")) {
+		/*
+		 * Workaround hot plugged in device not being
+		 * initialized by the ACPICA code.
+		 */
+		status = acpi_evaluate_object(sony_nc_acpi_handle,
+					      METHOD_NAME__INI,
+					      NULL, NULL);
+
+		if (ACPI_FAILURE(status)) {
+			pr_warn("Unable to call the _INI method.\n");
+			goto outplatform;
+		}
+	}
+
 	if (acpi_has_method(sony_nc_acpi_handle, "SN00")) {
 		dprintk("Doing SNC setup\n");
 		/* retrieve the available handles */
-- 
2.21.0

